File: php/src/Vbml.php
<?php

namespace Vestaboard\Vbml;

/**
 * Main VBML class.
 * Port of index.ts
 */
class Vbml
{
    private const BOARD_ROWS    = 6;
    private const BOARD_COLUMNS = 22;

    public static function parse(array $input): array
    {
        $height = $input['style']['height'] ?? self::BOARD_ROWS;
        $width  = $input['style']['width']  ?? self::BOARD_COLUMNS;
        $props  = $input['props'] ?? null;

        $emptyBoard = CreateEmptyBoard::create($height, $width);

        // Regular (non-absolute) components
        $components = array_values(array_filter(
            $input['components'],
            fn($c) => !isset($c['style']['absolutePosition']) && !isset($c['calendar'])
        ));
        $components = array_map(
            ParseComponent::parseComponent($height, $width, $props),
            $components
        );

        // Absolute components (non-calendar)
        $absoluteComponents = array_values(array_filter(
            $input['components'],
            fn($c) => !isset($c['calendar']) && isset($c['style']['absolutePosition'])
        ));
        $absoluteComponents = array_map(
            ParseComponent::parseAbsoluteComponent($height, $width, $props),
            $absoluteComponents
        );

        // Calendar components
        $calendarComponents = [];
        foreach ($input['components'] as $component) {
            if (!isset($component['calendar'])) {
                continue;
            }
            $cal = $component['calendar'];
            $x = $component['style']['absolutePosition']['x'] ?? 0;
            $calendar = Calendar::makeCalendar(
                $cal['month'],
                $cal['year'],
                $cal['days'] ?? [],
                $cal['defaultDayColor'] ?? null,
                $cal['hideSMTWTFS'] ?? false,
                $cal['hideDates'] ?? false,
                $cal['hideMonthYear'] ?? false
            );
            $calendarComponents[] = Calendar::parseCalendarComponent($calendar, $x);
        }

        return LayoutComponents::layout($emptyBoard, $components, $absoluteComponents, $calendarComponents);
    }

    public static function characterCodesToString(array $characters, array $options = []): string
    {
        return CharacterCodesToString::convert($characters, $options);
    }

    public static function characterCodesToAscii(array $characterCodes, bool $isWhite = false): string
    {
        return CharacterCodesToAscii::convert($characterCodes, $isWhite);
    }

    public static function copyCharacterCodes(array $characters): array
    {
        return CopyCharacterCodes::copy($characters);
    }

    public static function classic(string $text, array $options = []): array
    {
        return Classic::classic($text, $options);
    }

    public static function hasSpecialCharacters(?string $text): bool
    {
        return HasSpecialCharacters::check($text);
    }

    public static function sanitizeSpecialCharacters(string $text): string
    {
        return SanitizeSpecialCharacters::sanitize($text);
    }
}

File: python/pyvbml/vbml.py
"""VBML.

Port of Vestaboard/vbml/src/index.ts

Main entry point. Exposes the `vbml` namespace object and re-exports
all public types.
"""

from __future__ import annotations

from typing import cast

from .calendar import make_calendar
from .const import DIMENSIONS_FLAGSHIP
from .create_empty_board import create_empty_board
from .layout_components import layout_components
from .parse_calendar_component import parse_calendar_component
from .parse_component import parse_absolute_component, parse_component
from .types import IVBML, IVBMLComponent


def parse(input_data: IVBML) -> list[list[int]]:
    """Parse."""
    style = input_data.get("style") or {}
    height = style.get("height") or DIMENSIONS_FLAGSHIP[0]
    width = style.get("width") or DIMENSIONS_FLAGSHIP[1]
    props = input_data.get("props") or {}

    empty_board = create_empty_board(height, width)

    raw_components = input_data.get("components", [])

    def _has_abs(c: IVBMLComponent) -> bool:
        return bool((c.get("style") or {}).get("absolutePosition"))

    # Flow components (no absolute position, not a calendar)
    flow_components = [
        parse_component(height, width, props, c)
        for c in raw_components
        if not _has_abs(c) and "calendar" not in c
    ]

    # Absolute-positioned components (not calendars)
    absolute_components = [
        parse_absolute_component(height, width, props, c)
        for c in raw_components
        if _has_abs(c) and "calendar" not in c
    ]

    # Calendar components
    calendar_components = []
    for c in raw_components:
        if "calendar" not in c:
            continue
        cal_data = c["calendar"]
        x = cast(
            dict, cast(dict, c.get("style") or {}).get("absolutePosition") or {}
        ).get("x", 0)

        cal = make_calendar(
            month=cal_data["month"],
            year=cal_data["year"],
            default_day_color=cal_data.get("defaultDayColor"),
            highlighted_days=cal_data.get("days", {}),
            hide_day_of_week=cal_data.get("hideSMTWTFS", False),
            hide_dates=cal_data.get("hideDates", False),
            hide_month_year=cal_data.get("hideMonthYear", False),
        )
        calendar_components.append(parse_calendar_component(cal, x))

    return layout_components(
        empty_board, flow_components, absolute_components, calendar_components
    )
